core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -415,105 +433,107 @@ public HashDistributionTaskSource(
-                partitionSourceHandles.put(planNodeId, handle);
-            }
-
-            int taskPartitionId = 0;
-            ImmutableList.Builder<TaskDescriptor> partitionTasks = ImmutableList.builder();
-            for (Integer partition : union(partitionToSplitsMap.keySet(), partitionToExchangeSourceHandlesMap.keySet())) {
-                ListMultimap<PlanNodeId, Split> splits = partitionToSplitsMap.getOrDefault(partition, ImmutableListMultimap.of());
-                ListMultimap<PlanNodeId, ExchangeSourceHandle> exchangeSourceHandles = ImmutableListMultimap.<PlanNodeId, ExchangeSourceHandle>builder()
-                        .putAll(partitionToExchangeSourceHandlesMap.getOrDefault(partition, ImmutableMultimap.of()))
-                        // replicated exchange source will be added in postprocessTasks below
-                        .build();
-                Set<HostAddress> hostRequirement = partitionToNodeMap.get(partition);
-                partitionTasks.add(new TaskDescriptor(taskPartitionId++, splits, exchangeSourceHandles, new NodeRequirements(catalogRequirement, hostRequirement, taskMemory)));
-            }
 
-            List<TaskDescriptor> result = postprocessTasks(partitionTasks.build());
+                            int taskPartitionId = 0;
+                            ImmutableList.Builder<TaskDescriptor> partitionTasks = ImmutableList.builder();
+                            for (Integer partition : union(partitionToSplitsMap.keySet(), partitionToExchangeSourceHandlesMap.keySet())) {
+                                ListMultimap<PlanNodeId, Split> splits = partitionToSplitsMap.getOrDefault(partition, ImmutableListMultimap.of());
+                                ListMultimap<PlanNodeId, ExchangeSourceHandle> exchangeSourceHandles = ImmutableListMultimap.<PlanNodeId, ExchangeSourceHandle>builder()
+                                        .putAll(partitionToExchangeSourceHandlesMap.getOrDefault(partition, ImmutableMultimap.of()))
+                                        // replicated exchange source will be added in postprocessTasks below
+                                        .build();
+                                Set<HostAddress> hostRequirement = partitionToNodeMap.get(partition);
+                                partitionTasks.add(new TaskDescriptor(taskPartitionId++, splits, exchangeSourceHandles, new NodeRequirements(catalogRequirement, hostRequirement, taskMemory)));
+                            }
 
-            finished = true;
-            return result;
+                            List<TaskDescriptor> result = postprocessTasks(partitionTasks.build());
+                            finished = true;
+                            return result;
+                        }
+                    },
+                    directExecutor());
```

Comment:
better run this in the executor (since it is available) to avoid blocking split loading threads